### PR TITLE
[14.0][ADD] l10n_br_mdfe: add new module

### PR DESCRIPTION
**Pull request by marcelsavegnago**
This pull request introduces several new features and improvements related to the Brazilian electronic fiscal document system (MDFe). The changes include adding support for the MDFe document type, updating various models and views, and introducing new constants for MDFe handling.

### New Features and Enhancements:

* **MDFe Support:**
  - Added new MDFe document type in `l10n_br_fiscal/constants/fiscal.py` and `l10n_br_fiscal/data/l10n_br_fiscal.document.type.csv`. [[1]](diffhunk://#diff-02723dceadf60070a3ab92bc1526a98b06670b21f95d9587ebbe58ec6eaf94a8R336-R341) [[2]](diffhunk://#diff-3fda816ecdf056fdc52e25747eed2e84f802c2f9dbb9519f4a1fb3c588fa67d3R34)
  - Introduced MDFe document handling in `l10n_br_fiscal_edi/models/document_workflow.py`. [[1]](diffhunk://#diff-8bb36ca418552c9d0d99eb38bc389c348a2c84fb868bbfc19ad359a3e1e3b8b7R12) [[2]](diffhunk://#diff-8bb36ca418552c9d0d99eb38bc389c348a2c84fb868bbfc19ad359a3e1e3b8b7R240) [[3]](diffhunk://#diff-8bb36ca418552c9d0d99eb38bc389c348a2c84fb868bbfc19ad359a3e1e3b8b7R257-R259) [[4]](diffhunk://#diff-8bb36ca418552c9d0d99eb38bc389c348a2c84fb868bbfc19ad359a3e1e3b8b7R393-R401)
  - Added MDFe-specific constants in `l10n_br_mdfe/constants/mdfe.py`.

* **Model and View Updates:**
  - Updated `l10n_br_fiscal/models/document_mixin_fields.py` and `l10n_br_fiscal/models/document_related.py` to include new fields for MDFe. [[1]](diffhunk://#diff-3eb477272111a7819df116fab7a742d4c9c50bbdb37229b75099f3c7425b114bR470-R473) [[2]](diffhunk://#diff-6f5d2807baa2d51d5afc6abe81c85bcf2143f7f07b5c641d3e2a1febbd267d49R63-R71) [[3]](diffhunk://#diff-6f5d2807baa2d51d5afc6abe81c85bcf2143f7f07b5c641d3e2a1febbd267d49R104-R105)
  - Modified views in `l10n_br_fiscal/views/document_related_view.xml` and `l10n_br_fiscal/views/document_view.xml` to display new MDFe fields. [[1]](diffhunk://#diff-b812ca109c39260fed2101373f13fe31c7f6c2640050bfcd781d5da3b5ec9533R71-R80) [[2]](diffhunk://#diff-adc69877f2b558450c41e8898771f7c53c443ab67c5a6cd01409ddeb45fdd1a3R404)

* **Demo and Data Files:**
  - Added demo data for MDFe in `l10n_br_fiscal/demo/company_demo.xml`. [[1]](diffhunk://#diff-714d369661c92d5770bb3dc74c8a878671855f4184d32acf6ff581732af38bdbR165-R172) [[2]](diffhunk://#diff-714d369661c92d5770bb3dc74c8a878671855f4184d32acf6ff581732af38bdbR205-R212)
  - Added new MDFe operation records in `l10n_br_fiscal/data/operation_data.xml`.

* **Module Initialization and Manifest Updates:**
  - Created new `l10n_br_mdfe` module with its own `__init__.py`, `__manifest__.py`, and `README.rst`. [[1]](diffhunk://#diff-8a57f44549e357802c91c19b7acc18c6f8e8b2da292953a3f724ba3b86cb8e31R1-R2) [[2]](diffhunk://#diff-3642bbbff95946647b786dcc9b7559f9a698c4435cec191ac891ab1b9c8832e1R1-R48) [[3]](diffhunk://#diff-9ea342d2256543c996c4c40e982443cdb3b4fe117b3c2d693dcf73109a7b6e52R1-R123)
  - Updated version numbers in `l10n_br_fiscal/__manifest__.py` and `l10n_br_fiscal_edi/__manifest__.py`. [[1]](diffhunk://#diff-f7255250f2d05763c9ceeb4ae1d2a91bcedcf7d645a2a71dccbd02598b69118dL13-R13) [[2]](diffhunk://#diff-9d8607e5f0867bb6150bb8d2cbb446e032fd35e8897444af7ffab5cb3cdc17a7L12-R12)

These changes collectively enhance the system's capabilities to handle MDFe documents, ensuring compliance with Brazilian fiscal regulations.

### CODE REVIEW

File: l10n_br_mdfe_spec/models/spec_models.py [hunk 1]
```diff
@@ -7,13 +7,10 @@
 class MdfeSpecMixin(models.AbstractModel):
```
Comment (rvalyi) [inline l10n_br_mdfe_spec/models/spec_models.py:7]: pode renomear o arquivo spec_mixin.py ? Eu fiz isso no PR do multi-esquemas, reflete mais o que tem dentro.

Pois o spec.mixin.schema eh algo que extende TODOS modelos geridos pro XSD, enquanto que o SpecModel e StackedModel sao apenas pro objetos Odoo nativos que tem que ser extendido por algum mixin do esquemas. Ja tem uma certa complexidade entao melhor ajudar com o nome das coisas.

**Review by rvalyi — CHANGES_REQUESTED** [review]

File: l10n_br_mdfe/hooks.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Copyright (C) 2019-2020 - Raphael Valyi Akretion
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+from odoo import SUPERUSER_ID, api
+
+
+def post_init_hook(cr, registry):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    env["mdfe.30.tmdfe_infnfe"]._register_hook()
```
Comment (rvalyi) [inline l10n_br_mdfe/hooks.py]: Eu acho que pode tirar isso agora.

**Review by rvalyi — CHANGES_REQUESTED** [review]

File: l10n_br_mdfe_spec/models/spec_models.py (hunk 1 above)
Comment (marcelsavegnago) [reply l10n_br_mdfe_spec/models/spec_models.py:7]: Done

File: l10n_br_mdfe/hooks.py (hunk 2 above)
Comment (marcelsavegnago) [reply l10n_br_mdfe/hooks.py]: Done

**Comment by rvalyi** [comment]
@marcelsavegnago o _register_hook não foi chamado pros objetos da mdfe. Esse problema foi depois de vc remover o hook apenas? Ele deveria fazer sozinho mas talvez ele cai naquele problema que eu comentei mais cedo de precisar de um loop para cada esquema no #3431. 

Para as tabelas desses objetos existirem o _register_hook precisa ser chamado pelo menos uma vez para aquele esquema da mdfe. Normalmente ele deveria fazer sozinho, mas talvez caiu no caso que aquele mro() trouxe apenas as extensões da nfe e que ele parou apenas na primeira extensão "spec" da nfe no caso. Na pior se resolver de chamar o _register_hook em cima de um objeto da mdfe no hook.py, seria um workaround aceitável até fazer aquele loop no spec_driven_model.

**Review by rvalyi — CHANGES_REQUESTED** [review]
@marcelsavegnago eu tinha errado no summary do manifest do modulo l10n_br_mdfe_spec (ta como CTe). Pode aproveitar para alterar o summary por algo do tipo: "MDF-e abstract models generated by xsdata-odoo from the oficial xsd" ou algo do tipo que achar melhor por favor? (eu acabei de ajustar dessa forma no summary do l10n_br_nfe_spec em #3461)

**Review by rvalyi — CHANGES_REQUESTED** [review]
nos logs eu vi esse warning: "2024-10-21 13:08:15,686 535 WARNING odoo odoo.modules.registry: res.partner: inconsistent 'compute_sudo' for computed fields: mdfe30_CNPJ, mdfe30_idEstrangeiro, mdfe30_CPF, mdfe30_choice6, mdfe30_CEP, mdfe30_fone, mdfe30_IE, mdfe30_choice8, mdfe30_choice9, mdfe30_choice10, mdfe30_choice11, mdfe30_choice12 " Na NFe a gente conseguiu corrigir isso.

Tb é ruim usar esses padrões mdfe30_choiceXY, seria melhor botar um nome `mdfe30_choice_<alguma coisa explicita>` assim como foi feito nos campos choice da NFe.

**Comment by marcelsavegnago** [comment]
> nos logs eu vi esse warning: "2024-10-21 13:08:15,686 535 WARNING odoo odoo.modules.registry: res.partner: inconsistent 'compute_sudo' for computed fields: mdfe30_CNPJ, mdfe30_idEstrangeiro, mdfe30_CPF, mdfe30_choice6, mdfe30_CEP, mdfe30_fone, mdfe30_IE, mdfe30_choice8, mdfe30_choice9, mdfe30_choice10, mdfe30_choice11, mdfe30_choice12 " Na NFe a gente conseguiu corrigir isso.
> 
> Tb é ruim usar esses padrões mdfe30_choiceXY, seria melhor botar um nome `mdfe30_choice_<alguma coisa explicita>` assim como foi feito nos campos choice da NFe.

Done

**Comment by marcelsavegnago** [comment]
> @marcelsavegnago eu tinha errado no summary do manifest do modulo l10n_br_mdfe_spec (ta como CTe). Pode aproveitar para alterar o summary por algo do tipo: "MDF-e abstract models generated by xsdata-odoo from the oficial xsd" ou algo do tipo que achar melhor por favor? (eu acabei de ajustar dessa forma no summary do l10n_br_nfe_spec em #3461)

Done

**Comment by rvalyi** [comment]
@marcelsavegnago parabéns pelo trabalho!! Ficou OK para revisar?

File: l10n_br_mdfe/models/res_company.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        for rec in self:
+            if rec.partner_id.is_company:
+                rec.mdfe30_choice_emit = "mdfe30_CNPJ"
+            else:
+                rec.mdfe30_choice_emit = "mdfe30_CPF"
+
+    def _build_attr(self, node, fields, vals, path, attr):
+        if attr[0] == "enderEmit" and self.env.context.get("edoc_type") == "in":
+            # we don't want to try build a related partner_id for enderEmit
+            # when importing an NFe
```
Comment (rvalyi) [inline l10n_br_mdfe/models/res_company.py]: Corrigir NFe -> MDFe

**Comment by rvalyi** [comment]
começa a ficar top.



Do meu lado apenas falta acertar o script de migração do sub-PR https://github.com/OCA/l10n-brazil/pull/3446 para aprovar esse.

**Comment by marcelsavegnago** [comment]
@rvalyi vou mexer no script agora de manha..

**Comment by rvalyi** [comment]
rebase! por favor

**Comment by marcelsavegnago** [comment]
> rebase! por favor

feito :D

File: l10n_br_fiscal_edi/models/document_workflow.py [hunk 4]
```diff
@@ -385,3 +390,12 @@ def _action_document_correction(self):
...
             )
+
+    def _document_qrcode(self):
+        pass
+
+    def _processador(self):
+        pass
+
+    def _valida_xml(self, xml_file):
+        pass
```
Comment (rvalyi) [inline l10n_br_fiscal_edi/models/document_workflow.py:401]: Não pretendo blocar por isso, mas servem esses métodos? Se servir é melhor botar em inglês já...

**Review by rvalyi — APPROVED** [review]
Apenas uma observação, mas no geral eu acho OK.

**Comment by marcelsavegnago** [comment]
@rvalyi não investiguei mas ao alterar o document.supplement para spec ao invés de stack, deu erro no teste do nfce tentando buscar o Tmdfe no binding da nfe.. por ora voltei para stack

File: l10n_br_fiscal_edi/models/document_workflow.py (hunk 4 above)
Comment (marcelsavegnago) [reply l10n_br_fiscal_edi/models/document_workflow.py:401]: Renomeados

_valida_xml para _validate_xml
_processador para _edoc_processor

**Review by rvalyi — CHANGES_REQUESTED** [review]
para não dar problema no merge ainda teria que ajeitar o numero dos modulos. Ja que seria merge nobump, eu diria que pode dar uma incrementada menor em todos outros modulos e patch no l10n_br_focus.

**Comment by marcelsavegnago** [comment]
> para não dar problema no merge ainda teria que ajeitar o numero dos modulos. Ja que seria merge nobump, eu diria que pode dar uma incrementada menor em todos outros modulos e patch no l10n_br_focus.

Done

File: l10n_br_mdfe/demo/company_demo.xml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo noupdate="1">
+
+    <!-- Empresa Lucro Presumido -->
```
Comment (mileo) [inline l10n_br_mdfe/demo/company_demo.xml:4]: Acho que isso não deveria estar nesse módulo.

File: l10n_br_mdfe/models/document.py [hunk 6]
```diff
@@ -0,0 +1,910 @@
...
+    # MDF-e tag: tot
+    # Methods
+    ##########################
+
+    @api.depends(
+        "mdfe30_infMunDescarga.cte_ids",
+        "mdfe30_infMunDescarga.nfe_ids",
+        "mdfe30_infMunDescarga.mdfe_ids",
+    )
+    def _compute_tot(self):
```
Comment (mileo) [inline l10n_br_mdfe/models/document.py]: Pode melhorar o nome deste método?

File: l10n_br_mdfe/demo/company_demo.xml (hunk 5 above)
Comment (marcelsavegnago) [reply l10n_br_mdfe/demo/company_demo.xml:4]: De fato acho que não precisamos definir o processador. vou remover estes dados demo.. nao vejo necessidade nem de adicionar em outro módulo.. vc vê ?

Comment (marcelsavegnago) [reply l10n_br_mdfe/demo/company_demo.xml:4]: Seguinte.. vou deixar assim por enquanto. Está igual a NFE e podemos tratar isso posteriormente.

File: l10n_br_mdfe/models/document.py (hunk 6 above)
Comment (marcelsavegnago) [reply l10n_br_mdfe/models/document.py]: Este metodo calcula os campos da classe tot ("Totalizadores da carga transportada e seus documentos fiscais")

alguma sugestão ?

 _compute_tag_tot ?
_compute_transport_fiscal_totals
_compute_load_fiscal_totals
_compute_fiscal_transport_totals

Comment (marcelsavegnago) [reply l10n_br_mdfe/models/document.py]: Done 
https://github.com/OCA/l10n-brazil/pull/3445/commits/b4151ed795b4857d41537658cbe3e81099c100f4

**Review by rvalyi — APPROVED** [review]

**Review by kaynnan — APPROVED** [review]

**Comment by rvalyi** [comment]
ainda tem a gambiarra do document.suplement herdar do StackedModel, mas enfim o impacto não é muito grande, a gente consegue ajustar depois. Eu diria que a gente já poderia fazer o merge assim.

**Pull request merged**